### ISSUE TITLE
Incorrect type narrowing when using "x in y"

### ISSUE THREAD

**Issue by danpascu** [microsoft/pyright#10381]
**Describe the bug**
Using `key in mapping` test doesn't narrow the type correctly for user defined types.

I have a mapping from a key of type T to a value of type V.
I have a function receiving an argument that has a type that is a union between the key type T and other types (like `def get_link(key: T | U) -> Link`

If T and U are user defined types it fails to narrow it (see example 1 below).
If T and U are builtin types narrowing seems to work (see example 2 below).
If T and U are a mix of builtin and user defined types it fails (see example 3 below).

**Code or Screenshots**

```py
from typing import reveal_type

class NodeID(bytes): ...
class ResourceID(bytes): ...
class Link: ...

node_link_map: dict[NodeID, Link] = {}


def get_link(key: NodeID | ResourceID) -> Link:
    if key in node_link_map:
        reveal_type(key)           # Type of "key" is "NodeID | ResourceID" (expected NodeID)
        return node_link_map[key]  # Argument of type "NodeID | ResourceID" cannot be assigned to parameter "key" of type "NodeID" in function "__getitem__"
    raise ValueError


link_map2: dict[bytes, Link] = {}


def get_link_2(key: bytes | str) -> Link:
    if key in link_map2:
        reveal_type(key)       # Type of "key" is "bytes"
        return link_map2[key]  # No error
    raise ValueError


link_map3: dict[str, Link] = {}

def get_link_3(key: bytes | str | NodeID | ResourceID) -> Link:
    if key in link_map3:
        reveal_type(key)       # Type of "key" is "str | NodeID | ResourceID" (expected str)
        return link_map3[key]  # Argument of type "str | NodeID | ResourceID" cannot be assigned to parameter "key" of type "str" in function "__getitem__"
    raise ValueError
```

**VS Code extension or command-line**
Command line tool and language server in nvim. Version 1.1.400

**Comment by erictraut** [comment]
Static type checkers need to be conservative when applying type narrowing for the `x in y` type guard pattern. The `in` operator uses the `__eq__` method which can be overridden by a class to perform arbitrary logic. For that reason, type narrowing can be applied only under circumstances where the container type and the key type both have well-known characteristics. Pyright therefore restricts type narrowing to specific built-in classes that are commonly used in this pattern. When arbitrary user-defined classes are used, type checkers cannot safely make such assumptions.

Your code appears to contain a logic error. There's nothing that prevents a `NodeID` and a `ResourceID` from having the same underlying `bytes` representation. That means a simple equality check (and therefore an `in` operator) is insufficient to distinguish between a `NodeID` and a `ResourceID`.

```python
print(NodeID(b"123") == ResourceID(b"123"))  # True
```

Perhaps you have other constraints in your program that prevent this overlap from occurring (e.g. you have non-overlapping namespaces for the `bytes` representation of a `NodeID` versus a `ResourceID`), but a static type checker would have no way of knowing about such constraints.

You can eliminate the type error and document the assumption in your code by adding an assert in the `if` block:

```python
assert isinstance(key, NodeID)
```

**Comment by danpascu** [comment]
That might be true, but we are not evaluating runtime behaviour here. At runtime I could even add int or str keys in that mapping and none complains. But I expect a static type checker to know the type of the key from the declaration.

If I attempt to add a ResourceID to that mapping pyright rightly complains:

```py
node_link_map[ResourceID()] = Link()  # Error: Argument of type "ResourceID" cannot be assigned to parameter "key" of type "NodeID" in function "__setitem__"
```

Thus, if pyright knows that a ResourceID cannot be added to that mapping as a key, I also expect to understand that all keys in there are of type NodeID and thus can do the type narrowing.

Also in example 3 the type narrowing is completely off, even if we consider your argument. Why are NodeID and ResourceID that are subclasses of bytes considered along side str?

**Comment by erictraut** [comment]
Consider the following code, which is a simplified variant of your `get_link` function:

```python
def func(key: NodeID | ResourceID, map: dict[NodeID, Link]):
    if key in map:
        reveal_type(key)
        print("Is key a NodeID?")
        print(type(key)) # Prints "class <ResourceId>"

func(ResourceID(b"123"), {NodeID(b"123"): Link()})
```

If I understand you correctly, you are expecting pyright to narrow `key` to `NodeID` based on the `key in map` type guard. That's unsound, as you can see if you run the above code. I'm intentionally passing in a `ResourceID` to demonstrate that the `key in map` type guard does not guarantee that `key` is a `NodeID`.

In your third example, pyright is able to safely eliminate `bytes` from the original union because there are no `bytes` values that evaluate equal to a `str` value. One might make the argument that a type checker could safely eliminate values whose class derives from `bytes` as long as the class is marked `@final`, since it can be proven in that case that `bytes.__eq__` has not been overridden in that case. That seems like a very niche case though.

I'll note that mypy doesn't perform any narrowing in any of your examples, even in the second and third examples where pyright is smart enough to eliminate `str` and `bytes`, respectively.

This is all working as intended, so I don't consider this a bug.

**Comment by danpascu** [comment]
No, you're right. I noticed my error after posting my reply when I had a bit more time to mull over it. `x in y` is a runtime test, so a type checker cannot reliably derive static type information from it as it can return `True` just because `x` happens to be equal with something inside `y` despite being of a different type.

Thanks for pointing out the error I was making.

**Comment by danpascu** [comment]
I have one more question. I also tested `==` and as expected it behaves the same since it is based on `__eq__` as well.

But if I try the `is` operator it still doesn't narrow it down. Is this correct? And if so what is the explanation, since `is` basically guarantees that the 2 operands are the same object so they cannot be of 2 different types at the same time:

```py

def get_node(n1: NodeID | ResourceID, n2: NodeID) -> NodeID:
    if n1 is n2:
        reveal_type(n1)  # Type of "n1" is "NodeID | ResourceID" ???
        return n1  # Error
    ...
```

If n2 is known to be only NodeID and is the same object as n1 doesn't that also mean that n1 is now a NodeID only?

**Comment by erictraut** [comment]
> they cannot be of 2 different types at the same time

That's not strictly true since Python supports multiple inheritance, but your point stands if both classes were marked `@final`.

In any case, `x is y` not a supported type guard pattern. For a full list of type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards). Pyright does support `x is L` and `x is C` as type guards, since these are common usages. Neither pyright nor mypy support `x is y`. It's not a commonly-used idiom because it's generally not very safe — or rather, most developers don't understand how to use it safely.

Here's a technique that works:

```python
if isinstance(n1, NodeID) and n1 is n2:
```

**Comment by danpascu** [comment]
It would be helpful if the documentation mentioned that the `x in y` type guard only works to narrow `x` down if `x` is a builtin type with a well known behaviour and it does not work for user defined types.
